src/iocp/mod.rs

```diff
@@ -0,0 +1,821 @@
+            let iosb = &mut *self.as_ref().iosb().get();
+
+            // Check the status.
+            match iosb.Anonymous.Status {
+                STATUS_CANCELLED => {
+                    // Poll request was cancelled.
+                }
+
+                status if status < 0 => {
+                    // There was an error, so we signal both ends.
+                    event.readable = true;
+                    event.writable = true;
+                }
+
+                _ => {
+                    // Check in on the AFD data.
+                    let afd_data = &*afd_info.get();
+
+                    if afd_data.handle_count() >= 1 {
+                        let events = afd_data.events();
+
+                        // If we closed the socket, remove it from being polled.
+                        if events.contains(AfdPollMask::LOCAL_CLOSE) {
+                            let source = {
+                                let mut sources =
+                                    poller.sources.write().unwrap_or_else(|e| e.into_inner());
+                                sources.remove(&socket_state.socket).unwrap()
+                            };
+                            return source.begin_delete().map(|()| FeedEventResult::NoEvent);
+                        }
+
+                        // Report socket-related events.
+                        let (readable, writable) = afd_mask_to_event(events);
+                        event.readable = readable;
+                        event.writable = writable;
+                    }
+                }
+            }
+        }
+
+        // Filter out events that the user didn't ask for.
+        {
+            if !socket_state.interest.readable {
+                event.readable = false;
+            }
+
+            if !socket_state.interest.writable {
+                event.writable = false;
+            }
+        }
+
+        // If this event doesn't have anything that interests us, don't return or
+        // update the oneshot state.
+        let return_value = if event.readable || event.writable {
+            // If we are in oneshot mode, remove the interest.
+            if matches!(socket_state.mode, PollMode::Oneshot) {
+                socket_state.interest = Event::none(socket_state.interest.key);
+                socket_state.interest_error = false;
+            }
+
+            FeedEventResult::Event(event)
+        } else {
+            FeedEventResult::NoEvent
+        };
+
+        // Put ourselves in the update queue.
+        drop(socket_state);
+        poller.update_packet(self)?;
+
+        // Return the event.
+        Ok(return_value)
+    }
+
+    /// Begin deleting this socket.
+    fn begin_delete(self: Pin<Arc<Self>>) -> io::Result<()> {
+        // If we aren't already being deleted, start deleting.
+        let mut socket = self
+            .as_ref()
+            .socket_state()
+            .expect("can't delete notification packet");
```

Comment:
this message could be confusing, as it is missing context that is also not obvious from the function name. A clear indication that the notification packet couldn't be deleted because it wasn't a socket might be a good idea.